system/CodeIgniter.php

```diff
@@ -1006,21 +1010,9 @@ protected function display404errors(PageNotFoundException $e)
      */
     protected function gatherOutput(?Cache $cacheConfig = null, $returned = null)
     {
-        $this->output = ob_get_contents();
-        // If buffering is not null.
-        // Clean (erase) the output buffer and turn off output buffering
-        if (ob_get_length()) {
-            ob_end_clean();
-        }
+        $this->output = $this->outputBufferingEnd();
 
         if ($returned instanceof DownloadResponse) {
-            // Turn off output buffering completely, even if php.ini output_buffering is not off
-            if (ENVIRONMENT !== 'testing') {
-                while (ob_get_level() > 0) {
-                    ob_end_clean();
-                }
-            }
-
```

Comment:
Revert these lines. It is intentional.
See #5623

Reply:
Before checking the response type, the ``outputBufferingEnd()`` method is called which contains the same code.
```php
        while (ob_get_level() > $this->bufferLevel) {
            $buffer .= ob_get_contents();
            ob_end_clean();
        }
```
In a default application, ``$this->bufferLevel`` will always be ``0``.

Reply:
```diff
--- a/system/CodeIgniter.php
+++ b/system/CodeIgniter.php
@@ -1146,6 +1146,7 @@ class CodeIgniter
     protected function outputBufferingStart(): void
     {
         $this->bufferLevel = ob_get_level();
+        dd($this->bufferLevel);
         ob_start();
     }
 
```
```
$this->bufferLevel integer 1
⧉ Called from .../system/CodeIgniter.php:1149 [dd()]
```

Reply:
I was wrong.
But can we move this code into the DownloadResponse class and call it directly in the ``send`` method?
Like this
```php
// DownloadResponse::class
    public function send()
    {
        if (ENVIRONMENT !== 'testing') {
            while (ob_get_level() > 0) {
                ob_end_clean();
            }
        }
        $this->buildHeaders();
        $this->sendHeaders();
        $this->sendBody();

        return null;
    }
```

Reply:
Okay, it might be better.
Please keep the comment `// Turn off output buffering completely, even if php.ini output_buffering is not off`.